src/bitbox02.js

```diff
@@ -86,25 +93,93 @@ const setOutputDefaults = outputs =>  {
 }
 
 export class BitBox02API {
+    /**
+     * @param devicePath See `getDevicePath()`.
+    */
     constructor(devicePath)  {
         this.devicePath = devicePath;
         this.opened = false;
-        this.socket = null;
+        // connection is an object with three keys once the connection is established:
+        // onWrite(bytes): send bytes
+        // close():  close the connection
+        // valid(): bool - is the connection still alive?
+        this.connection = null;
+        this.onCloseCb = null;
+
+        if (navigator.hid) {
+            navigator.hid.addEventListener("disconnect", () => {
+                if (this.onCloseCb) {
+                    this.onCloseCb();
+                }
+            });
+        }
     }
 
-    connectWebsocket() {
+    connectWebsocket = onMessageCb => {
         const socket = new WebSocket("ws://127.0.0.1:8178/api/v1/socket/" + this.devicePath);
         return new Promise((resolve, reject) => {
             socket.binaryType = 'arraybuffer';
+            socket.onmessage = event => { onMessageCb(new Uint8Array(event.data)); };
+            socket.onclose = event => {
+                if (this.onCloseCb) {
+                    this.onCloseCb();
+                }
+            };
             socket.onopen = function (event) {
-                resolve(socket);
+                resolve({
+                    onWrite: bytes => {
+                        if (socket.readyState != WebSocket.OPEN) {
+                            console.log("Error, trying to write to closed socket");
```

Comment:
`console.error("attempted write to a closed socket")`? same in connectWebHID.

Reply:
done